x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialRelatesFunction.java

```diff
@@ -188,4 +194,94 @@ protected boolean pointRelatesGeometry(Point point, Component2D component2D) {
             }
         }
     }
+
+    /**
+     * When dealing with ST_CONTAINS and ST_WITHIN we need to pre-combine the field geometries for multi-values in order
+     * to perform the relationship check. This means instead of relying on the generated evaluators to iterate over all
+     * values in a multi-value field, the entire block is passed into the spatial function, and we combine the values into
+     * a geometry collection or multipoint.
+     */
+    protected interface MultiValuesCombiner {
+        Geometry combined();
+    }
+
+    /**
+     * Values read from source will be encoded as WKB in BytesRefBlock. The block contains multiple rows, and within
+     * each row multiple values, so we need to efficiently iterate over only the values required for the requested row.
+     * This class works for point and shape fields, because both are extracted into the same block encoding.
+     * However, we do detect if all values in the field are actually points and create a MultiPoint instead of a GeometryCollection.
+     */
+    protected static class MultiValuesBytesRef implements MultiValuesCombiner {
```

Comment:
I wonder if this abstraction is really needed. The methods in this class can be implemented as helper methods in this class?

Reply:
It is possible, but makes the code more complex. We have 4 methods that take the interface type `MultiValuesCombiner`), and these would need to be doubled to 8 methods (for LongBlock and BytesRefBlock versions) and each would need to add additional position parameters. I think the added complexity is not worth it.